lib/MapCreator/CaveCreator.js

```diff
@@ -0,0 +1,154 @@
+var Direction = require('../GameObjects/Direction');
+var Map = require('../Map');
+var MapUtils = require('../Utils/MapUtils');
+var MapLocation = require('../GameObjects/MapLocation');
+var MapOptions = require('./MapOptions');
+var TerrainType = require('../GameObjects/TerrainType');
+
+function cave(height, width, mapOptions) {
+    var start, exit, map, i, j;
+    var primaryEnemy = mapOptions[MapOptions.PRIMARY_ENEMY];
+    var secondaryEnemy = mapOptions[MapOptions.SECONDARY_ENEMY];
+    var rareEnemy = mapOptions[MapOptions.RARE_ENEMY];
+    var boss = mapOptions[MapOptions.BOSS_ENEMY];
+    var fill = mapOptions[MapOptions.TERRAIN_FILL] || 1;
+    var roundLimit = mapOptions[MapOptions.ROUND_LIMIT] || 1000;
+
+    start = new MapLocation(1, 1);
+    exit = new MapLocation(width - 2, height - 2);
+
+    // Approximately the % of open space, must be in [.1,.6]
+    map = new Map(height, width, start, exit, roundLimit, TerrainType.ROCK);
+    map.fillWithTiles(TerrainType.ROCK);
+
+    // Choose a few centers for large openings in the cave.
+    var centers = [start, exit];
+    i = 0;
+    var idx = 0;
+    while(i < fill * height * width / 25 + Math.random() * 2) {
+        idx++;
+        if (idx > 200) {
+            break;
+        }
+        var newCenter = new MapLocation(Math.floor(Math.random() * width), Math.floor(Math.random() * height));
+        var validLoc = true;
+        for (j = 0; j < centers.length; j++) {
+            if (centers[j].distanceSquaredTo(newCenter) < 25) {
+                validLoc = false;
+            }
+        }
+        if (newCenter.x <= 2 || newCenter.x >= width - 3 || newCenter.y <= 2 || newCenter.y >= height - 3) {
+            validLoc = false;
+        }
+        if (validLoc) {
+            i++;
+            centers.push(newCenter);
+        }
+    }
+
+
+    // Grow large rooms randomly around centers
+    map.clearArea(start, start.add(Direction.SOUTH_EAST), TerrainType.CEMENT);
+    map.clearArea(exit.add(Direction.NORTH_WEST), exit, TerrainType.CEMENT);
+    for (i = 2; i < centers.length; i++) {
```

Comment:
might work to pick two squares in opposite directions and build a circle around each